### ISSUE TITLE
after_commit isn't triggered on destroy when mutual dependent: :destroy + non-standard :foreign_key

### ISSUE THREAD

**Issue by Kukunin** [rails/rails#36335]
### Steps to reproduce
```ruby
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails', '5.2.3'
  gem 'sqlite3', '~> 1.3.6'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :directs do |t|
    t.bigint :user_id, null: false
    t.timestamps
  end

  create_table :aliases do |t|
    t.bigint :customer_id, null: false
    t.timestamps
  end

  create_table :users do |t|
    t.string :name
    t.timestamps
  end

  create_table :logs do |t|
    t.timestamps
  end
end

class Log < ActiveRecord::Base; end

class Direct < ActiveRecord::Base
  belongs_to :user, dependent: :destroy
end

class Alias < ActiveRecord::Base
  belongs_to :customer, class_name: 'User', dependent: :destroy
end

class User < ActiveRecord::Base
  has_one :direct, dependent: :destroy
  has_one :alias, foreign_key: :customer_id, dependent: :destroy

  after_commit on: :destroy do
    puts 'callback'
    Log.create!
  end
end

class BugTest < Minitest::Test
  def test_after_commit_on_destroy_with_no_dependencies
    user = User.create!(name: 'User')

    log_count = Log.count
    user.destroy!
    assert_equal(log_count + 1, Log.count)
  end

  def test_after_commit_on_destroy_with_direct
    user = User.create!(name: 'User')
    Direct.create!(user: user)

    log_count = Log.count
    user.destroy!
    assert_equal(log_count + 1, Log.count)
  end

  def test_after_commit_on_destroy_with_alias
    user = User.create!(name: 'User')
    Alias.create!(customer: user)

    log_count = Log.count
    user.destroy!
    assert_equal(log_count + 1, Log.count)
  end
end
```

### Expected behavior
All tests should pass, `after_commit on: :destroy` should be called after destroy

### Actual behavior
`test_after_commit_on_destroy_with_alias` test fails. 
If there is an association with mutual 'dependent: :destroy`, the callback isn't fired

### System configuration
**Rails version**:
5.2.3, going to try with master

**Ruby version**:
2.5.3p105

**Comment by Kukunin** [comment]
Can't reproduce on `master`. Can it be backported somehow to 5.2.x?

**Comment by Kukunin** [comment]
I did `git bisect`, and found the commit it was fixed in `master`: 9252da96597fbffe2246704556524c4804239552

```
commit 9252da96597fbffe2246704556524c4804239552
Author: Ryuta Kamizono <kamipo@gmail.com>
Date:   Wed Apr 10 17:49:42 2019 +0900

    Don't call after_commit callbacks despite a record isn't saved
    
    Regardless of a record isn't saved (e.g. validation is failed),
    `after_commit` / `after_rollback` callbacks are invoked for now.
    
    To fix the issue, this adds a record to the current transaction only
    when a record is actually saved.
    
    Fixes #29747.
    Closes #29833.

:040000 040000 9f5025658695cf448d0aeb231673dcd9968dfe4c c7648e47dc2a2bec279b73fb76375962279f3309 M	activerecord
```
